source/simulator/solver_schemes.cc

```diff
@@ -449,15 +450,22 @@ namespace aspect
 
     assemble_stokes_system();
 
+    // recompute rhs
+    dcr.velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+    dcr.pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+    dcr.residual = std::sqrt(dcr.velocity_residual * dcr.velocity_residual + dcr.pressure_residual * dcr.pressure_residual);
+
+    // Test whether the rhs has dropped so much that we can assume that the iteration is done.
+    if (dcr.residual < dcr.residual_old * 1e-8)
+      {
+        return;
```

Comment:
Could you output some status message here? Say
```suggestion
        pcout << "    Nonlinear residual is small; skipping Stokes solve" << std::endl;
        return;
```

Reply:
I don't think that your new message text reflects what is happening well. You are saying that "it has been very large", but that's not the point -- it's just been the regular nonlinear residual. The point is that it is vastly smaller this time around, and that that's likely because the model is (more or less) a linear model that is adequately solved in one iteration, and you're now in the next iteration. I think that my text reflected that better: it's about being small this time; it may or may not have been large last time around, we really neither know nor care.

Reply:
Related, what you print in parentheses is not actually the old residual that you reference when you say "has been very large"), but the ratio. It's confusing to see a statement "has been very large (1.234e-13)" I think :-)

Reply:
hmm, isn't the point that the reduction in the residual has been very large? Do you have a better way of phrasing it? Should I just state that the reduction in the residual has been very large, and/or do you want me to show a different number?

Reply:
I think I misread the text. I though you had said "Nonlinear residual has been very large", but you're saying "Nonlinear residual *reduction* has been very large. Yes, now it makes sense to me.